src/freenet/pluginmanager/PluginInfoWrapper.java

```diff
@@ -142,9 +142,15 @@ public void startShutdownPlugin(PluginManager manager, boolean reloading) {
 		unregister(manager, reloading);
 		// TODO add a timeout for plug.terminate() too
 		System.out.println("Terminating plugin "+this.getFilename());
+
+		// set the plugin’s class loader as context class loader
+		ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
+		Thread.currentThread().setContextClassLoader(plug.getClass().getClassLoader());
+
 		try {
 			plug.terminate();
 		} catch (Throwable t) {
+			Thread.currentThread().setContextClassLoader(originalClassLoader);
```

Comment:
Wouldn't need to restore the original class loader on a successful termination? (maybe add a `finally` block?)

Reply:
That’s debatable… :) 
* This thread will also be used to terminate all other plugins so the context class loader will be overwritten again, anyway.
* In theory, most things shouldn’t even use the class loader.

But yeah, we should be consistent here, either reset it always or never but only in case of an exception. :)